django/test/testcases.py

```diff
@@ -90,6 +91,15 @@ def to_list(value):
     return value
 
 
+def is_pickable(obj):
+    """Returns true if the object can be dumped and loaded through the pickle module."""
+    try:
+        pickle.loads(pickle.dumps(obj))
+    except (AttributeError, TypeError):
+        return False
+    return True
```

Comment:
I'm not sure if it's an acceptable solution. I can imagine that `dumps`/`loads` round-trip for every attribute in every test will cause a big performance regression.

Reply:
The options I'd see to avoid/reduce the effect this could either:
1. Hard-code the list of attributes that will be pickled. But that's hard to maintain and we'll forget attributes.
2. Only run this function on a subset of attributes (those not starting with `_` for instance). But since the TestCase attributes can contain custom ones, we won't be able to cover them all.

But do we execute it that often?
We only call it when calling `SimpleTestCase.__getstate__`, which should only be called when we have to pickle test case.

Are all tests/subtests always pickled on parallel runs, or only when they fail? 🤔 I'm unsure of the multiprocessing behaviour of tests.
If the logic is only executed when the tests fail, this piece of code should be rather safe in terms of performances I reckon.

Reply:
Let's ask Adam for opinion (\cc @adamchainz).

Reply:
@David-Wobrock Unfortunately, it's called many times when run in parallel. For example:
```
python -m cProfile ./runtests.py queries
...
     2778    0.003    0.000    0.019    0.000 testcases.py:96(is_pickable)
```
No parallel, no `is_pickable()` calls:
```
python -m cProfile ./runtests.py queries --parallel=1
```

Maybe this is something we could avoid :thinking:

Reply:
FWIW there must be a way to create an `pickle.Unpickler` instance so it drops the keys in a `dict` that cannot be unpickled to avoid multiple `loads` roundtrips. Given we already venture in peeking at internals for `unittest` I figured it was at least worth mentioning.